File: src/vector.lua
local vector = {}

function vector.sign(x)
	if x<0 then
		return -1
	elseif x>0 then
		return 1
	else
		return 0
	end
end

function vector.magnitude(x, y)
	local r = math.sqrt(x^2 + y^2)
	return r
end

function vector.angle(x, y)
	local t = math.atan2(y, x)
	return t
end

function vector.components(r, angle)
	local x = r * math.cos(angle)
	local y = r * math.sin(angle)
	return x, y
end

function vector.rotate(x, y, angle)
	local r = vector.magnitude(x, y)
	local t = vector.angle(x, y)
	return vector.components(r, t + angle)
end

function vector.add(a, b)
	local l = {}
	l[1], l[2] = vector.rotate(b[1], b[2], a[3])
	l[3] = b[3]
	l[4], l[5] = vector.rotate(b[4] or 0, b[5] or 0, a[3])
	l[6] = b[6] or 0

	for i = 1, 6 do
		l[i] = l[i] + a[i]
	end

	return l
end

return vector

File: src/vector.rs
use mlua::prelude::LuaTable;
use mlua::{Lua, Result};

use crate::world::types::Location;

pub fn sign(x: f64) -> f64 {
    if x < 0.0 {
        -1.0
    } else if x > 0.0 {
        1.0
    } else {
        0.0
    }
}

pub fn magnitude(x: f64, y: f64) -> f64 {
    f64::sqrt(x * x + y * y)
}

pub fn angle(x: f64, y: f64) -> f64 {
    f64::atan2(y, x)
}

pub fn components(r: f64, angle: f64) -> (f64, f64) {
    let x: f64 = r * f64::cos(angle);
    let y: f64 = r * f64::sin(angle);

    (x, y)
}

pub fn rotate(x: f64, y: f64, angle: f64) -> (f64, f64) {
    let r: f64 = magnitude(x, y);
    let t: f64 = self::angle(x, y);
    components(r, t + angle)
}

pub fn add(a: Location, b: Location) -> Location {
    let mut l = Location {
        x: 0.0,
        y: 0.0,
        angle: 0.0,
        x_velocity: 0.0,
        y_velocity: 0.0,
        angle_velocity: 0.0,
    };
    (l.x, l.y) = rotate(b.x, b.y, b.angle);
    l.angle = b.angle;

    (l.x_velocity, l.y_velocity) = rotate(b.x_velocity, b.y_velocity, b.angle_velocity);
    l.angle_velocity = b.angle_velocity;

    l.x += a.x;
    l.y += a.y;
    l.angle += a.angle;
    l.x_velocity += a.x_velocity;
    l.y_velocity += a.y_velocity;
    l.angle_velocity += a.angle_velocity;

    l
}

pub fn lua_module(lua: &Lua) -> Result<LuaTable> {
    let exports = lua.create_table()?;

    exports.set(
        "sign",
        lua.create_function(|_lua: &Lua, x: f64| Ok(sign(x)))?,
    )?;
    exports.set(
        "magnitude",
        lua.create_function(|_lua: &Lua, (x, y): (f64, f64)| Ok(magnitude(x, y)))?,
    )?;
    exports.set(
        "angle",
        lua.create_function(|_lua: &Lua, (x, y): (f64, f64)| Ok(angle(x, y)))?,
    )?;
    exports.set(
        "components",
        lua.create_function(|_lua: &Lua, (r, angle): (f64, f64)| Ok(components(r, angle)))?,
    )?;
    exports.set(
        "rotate",
        lua.create_function(|_lua: &Lua, (x, y, angle): (f64, f64, f64)| Ok(rotate(x, y, angle)))?,
    )?;
    exports.set(
        "add",
        lua.create_function(|_lua: &Lua, (a, b): (Location, Location)| Ok(add(a, b)))?,
    )?;

    Ok(exports)
}
